javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/PropagatorsInitializer.java

```diff
@@ -82,73 +102,20 @@ public static void initializePropagators(List<String> propagatorIds) {
       }
     }
     if (propagators.size() > 1) {
-      textMapPropagators.add(new MultiPropagator(propagators));
+      textMapPropagators.addAll(propagators);
     } else if (propagators.size() == 1) {
       textMapPropagators.add(propagators.get(0));
     }
-    // Register it in the global propagators:
-    OpenTelemetry.setGlobalPropagators(
-        ContextPropagators.create(TextMapPropagator.composite(textMapPropagators)));
+    return createPropagatorsRemovingNoops(textMapPropagators);
   }
 
-  private static TracerProvider unobfuscate(TracerProvider tracerProvider) {
-    if (tracerProvider.getClass().getName().endsWith("TracerSdkProvider")) {
-      return tracerProvider;
-    }
-    try {
-      Method unobfuscate = tracerProvider.getClass().getDeclaredMethod("unobfuscate");
-      unobfuscate.setAccessible(true);
-      return (TracerProvider) unobfuscate.invoke(tracerProvider);
-    } catch (Throwable t) {
-      return tracerProvider;
-    }
-  }
-
-  static class MultiPropagator implements TextMapPropagator {
-    private final List<Propagator> propagators;
-    private final List<String> fields;
-
-    private MultiPropagator(List<Propagator> propagators) {
-      this.propagators = propagators;
-
-      Set<String> fields = new LinkedHashSet<>();
-      for (Propagator propagator : propagators) {
-        fields.addAll(propagator.delegate.fields());
-      }
-      this.fields = new ArrayList<>(fields);
-    }
-
-    @Override
-    public List<String> fields() {
-      return fields;
-    }
-
-    @Override
-    public <C> void inject(Context context, C carrier, Setter<C> setter) {
-      for (Propagator propagator : propagators) {
-        propagator.inject(context, carrier, setter);
-      }
-    }
-
-    @Override
-    public <C> Context extract(Context context, C carrier, Getter<C> getter) {
-      boolean spanContextExtracted = false;
-      for (int i = propagators.size() - 1; i >= 0; i--) {
-        Propagator propagator = propagators.get(i);
-        if (!propagator.alwaysRun() && spanContextExtracted) {
-          continue;
-        }
-        context = propagator.extract(context, carrier, getter);
-        if (!spanContextExtracted) {
-          spanContextExtracted = isSpanContextExtracted(context);
-        }
-      }
-      return context;
-    }
-
-    private static boolean isSpanContextExtracted(Context context) {
-      return Span.fromContextOrNull(context) != null;
-    }
+  private static ContextPropagators createPropagatorsRemovingNoops(
+      List<TextMapPropagator> textMapPropagators) {
+    return ContextPropagators.create(
+        TextMapPropagator.composite(
+            textMapPropagators.stream()
+                .filter(propagator -> propagator != TextMapPropagator.noop())
+                .collect(toSet())));
```

Comment:
Just a thought: does the order of propagators matter here?

Reply:
It shouldn't...I mean, I don't think so.

There's also [a recent spec change](https://github.com/open-telemetry/opentelemetry-specification/pull/1316/files#diff-adf3b6d20af39e3109b1907f4b1f356e90912b893693a308955dea13dd97612f) that does require this deduplication.  I would expect that if order mattered that it would also have to call out how to deduplicate (first wins vs. last wins, etc).

Reply:
It's not totally clear from the spec, but I think order may be relevant, e.g.

> The resulting composite Propagator will invoke the Propagators, Injectors, or Extractors, in the order they were specified.

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#composite-propagator

Reply:
It's an interesting idea for sure.  There seems to be a lack of clarity on _how_ to deduplicate while maintaining order (first wins, last wins, etc).  The spec thinks it's important, but I'm not sure _why_.  Shouldn't matter, right?

This is further complicated by the SPI.  So if the agent config has `[b3, jaeger, b3]` and the spi has `[b3]`, what do we think the final deduplicated order should be?

Reply:
I think ordering is important on extraction, in case multiple propagators set the parent (propagated) span into the context, e.g. if a request has both jaeger and w3c headers.

It could potentially matter on injection) in case multiple propagators set the same http headers, though I think most propagators have their own set of http headers, so this seems less important of a concern.

> This is further complicated by the SPI. So if the agent config has [b3, jaeger, b3] and the spi has [b3], what do we think the final deduplicated order should be?

Which SPI are you thinking of?

Reply:
Seems edge casey, but I agree with the two possibilities you've outlined.

The SPI I was thinking about (and the entire motivation in submitting this PR) is along the TracerCustomizer, wherein it can set up its own propagators (like in the `DemoTracerCustomizer` example).  

If that happens, we will need to define if those come before or after the other agent propagators.  I would think that a custom one should come first....but I dunno.

Reply:
In `DemoTracerCustomizer`, can you programmatically order it as you wish? e.g. to put your custom propagator first, something like

```
GlobalOpenTelemetry.setPropagators(
    ContextPropagators.create(
        TextMapPropagator.composite(
            new DemoPropagator(), GlobalOpenTelemetry.getPropagators().getTextMapPropagator()))
```

Reply:
oh, I just saw https://github.com/open-telemetry/opentelemetry-java/pull/2475, @anuraaga is planning to bring over the new SDK auto-configuration to this repo after Java SDK 0.14.0, so probably we will inherit this SPI (and it's associated ordering)

Reply:
although, from vendor perspective, I'd think they would want to provide a "named" propagator, and have that configurable as part of `otel.propagators`, e.g. I would want to add an `applicationinsights` named propagator, and I'd want to override the default `otel.propagators` to be `tracecontext,baggage,applicationinsights`

Reply:
I haven't given it much thought, but I like the idea of symbolic names to help make things easier.  Seems outside the scope of this PR and #1681 though, doesn't it? 

Do we think there is a path forward with the approach in this PR (combine the preconfigured list ahead of the agent's configured list and dedupe), or do you think there's too many unknowns and other related changes for this to be viable?

Reply:
For the propagator provider, while haven't done so yet I'm intending to make `ConfigProperties` public so that can be used to recognize the `otel.propagators` property without having to force interaction between the SPI and the property It makes me realize the provider has to support not returning a propagator :)

For ordering, I copied over this comment we've had since the beginning

`Invocation order of {@code TextMapPropagator#inject()} and {@code TextMapPropagator#extract()} for registered trace propagators is undefined.`
https://github.com/open-telemetry/opentelemetry-java/blob/master/api/context/src/main/java/io/opentelemetry/context/propagation/TextMapPropagator.java#L50

Spec doesn't talk about order

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/context/api-propagators.md#composite-propagator

So I don't think the issue of order is something we need to worry about on this repo as the SDK doesn't guarantee any order anyways :) If that seems like a bad idea, we would probably want to confirm with a spec issue.

Reply:
@anuraaga the spec section you linked to does talk about ordering:

> The resulting composite Propagator will invoke the Propagators, Injectors, or Extractors, in the order they were specified

Reply:
Oops missed it :) Going to have to update the SDK javadoc. Do you think not enabling a provided propagator by default and requiring a `name` is a better approach to handle the ordering?

Reply:
> Do you think not enabling a provided propagator by default and requiring a `name` is a better approach to handle the ordering?

For the javaagent this seems the least confusing. Custom propagators then get treated the same as built-in propagators from configuration perspective.

Maybe the auto-configuration module needs a PropertySource SPI also, so that auto-configuration users can change the default `otel.propagators` value to include their custom propagator(s), same as javaagent distributions?

Reply:
@trask I'm leaning towards having two, one that is automatically enabled (most common in apps I think) and one that registers an available propagator by name. Just wondering, is it common in these custom distros to want to allow a user to disable the custom propagator using `otel.propagators`?

Reply:
Application Insights moved to W3C a while ago, but we still provide legacy propagation for backwards compatibility. Our default is to propagate both, but it's nice for users who don't need the backwards compatibility to be able to disable the legacy format (reduce wire size a little).

I could see this being a similar predicament for other vendors/users as they migrate away from legacy propagation formats.

Reply:
@breedx-splk thanks for instigating this discussion, it was very timely and led to good last minute changes in https://github.com/open-telemetry/opentelemetry-java/pull/2475 which I think will address #1681 once @anuraaga brings in the new autoconfigure module

Reply:
Sure thing!  Glad there was something good that fell out of it.  Whatever we end up with, hopefully it will have test coverage.  :)

Going to close since it sounds like there's a lot of momentum on another superseding path.